ktor-http/ktor-http-cio/src/io/ktor/http/cio/internals/WeakTimeoutQueue.kt

```diff
@@ -74,34 +74,29 @@ class WeakTimeoutQueue(
     suspend fun <T> withTimeout(block: suspend CoroutineScope.() -> T): T {
         return suspendCoroutineUninterceptedOrReturn { rawContinuation ->
             val continuation = rawContinuation.intercepted()
+
             val wrapped = WeakTimeoutCoroutine(continuation.context, continuation)
             val handle = register(wrapped)
-            wrapped.disposeOnCompletion(handle)
+            wrapped.invokeOnCompletion(handle)
 
             val result = try {
-                if (wrapped.isCancelled) throw wrapped.getCancellationException()
-                block.startCoroutineUninterceptedOrReturn(receiver = wrapped, completion = wrapped)
+                if (wrapped.isCancelled) COROUTINE_SUSPENDED
+                else block.startCoroutineUninterceptedOrReturn(receiver = wrapped, completion = wrapped)
             } catch (t: Throwable) {
-                CompletedExceptionally(t)
+                if (wrapped.tryComplete()) {
+                    handle.dispose()
+                    throw t
+                }
+                else COROUTINE_SUSPENDED
```

Comment:
This code path is not covered by tests and it's unclear why is it here.
Is it protection against concurrent cancel + completion race? If so, it should be outcome of the coroutine, not `COROUTINE_SUSPENDED `

Reply:
Similar to the pervious case, if it is already cancelled concurrently we need to return `COROUTINE_SUSPENDED` since the continuation is already resumed. Passing it to `result` will cause `suspendCoroutineUninterceptedOrReturn` return `COROUTINE_SUSPENDED` as well. However with the propsed optimization it becomes impossible so I'll simplify it.